File: Codes/jxdeng3264/000-100/002.Add-Two-Numbers/Solution.cpp
//ID:2.Add Two Numbers
//Author:jxdeng3989
//Alg&Ds: linklist math
//Time:2018-10-20
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode *res = NULL;
        ListNode *curnode = NULL;
        int carry = 0;
        int v1 = 0, v2 = 0;
        while( l1!=NULL || l2!=NULL)
        {
            if (l1!=NULL)
                v1 = l1->val;
            else
                v1 = 0;
            if (l2!=NULL)
                v2 = l2->val;
            else
                v2 = 0;

            ListNode *newnode = new ListNode((v1+v2+carry)%10);
            carry = (v1+v2+carry)/10;

            if(res==NULL)
                res = newnode;

            if(curnode==NULL)
                curnode = newnode;
            else
            {
                curnode->next = newnode;
                curnode = newnode;
            }
            if (l1!=NULL)
                l1 = l1->next;
            if(l2!=NULL)
                l2 = l2->next;
        }
        if( carry!=0)
        {
            ListNode *newnode = new ListNode(carry);
            curnode->next = newnode;
        }
        return res;
    }
};

File: Codes/jxdeng3264/000-100/002.Add-Two-Numbers/Solution.py
#!/usr/bin/env python
# encoding: utf-8
# author: pyclearl


# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None
# leetcode这里要注释掉才能AC


class Solution(object):
    def __init__(self):
        self.carry = 0

    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        p, q = l1, l2
        result = curr = ListNode(0)
        while p or q:
            val_p = p.val if p else 0
            val_q = q.val if q else 0
            _sum = self.carry + val_p + val_q
            self.carry = _sum / 10
            curr.next = ListNode(_sum % 10)
            curr = curr.next
            if p:
                p = p.next
            if q:
                q = q.next
        if self.carry > 0:
            curr.next = ListNode(self.carry)
        return result.next


if __name__ == "__main__":
    pass
